### PR TITLE
Add `ttt_map_settings`, `ttt_credit_adjust`, `ttt_feed_entry`, `ttt_force_win`, `ttt_logic_assigned`, `ttt_role_button` and round change event

### PR DESCRIPTION
**Pull request by Fires1**
Ready for review.

- Added `tttreborn.round.changed` event
- - Passes the `BaseRound round` parameter of `Game.ForceRoundChange` when fired.
- Added `ttt_map_settings` entity
- - Ironically has no settings at the moment but provides Outputs for round changes.
- - Can later be used for player models and such.
- Added `ttt_credit_adjust` entity
- - Identical logically to credit adjust entity from Gmod TTT. Removes/adds credits. 
- - Does not check if role is traitor or detective, can innocents hold credits? I do not know.
- Added `ttt_feed_entry`
- - Adds text to upper right feed. Can be sent to different people. Very clean.
- Added `ttt_force_win`
- - Forces a win for a certain team, can be the activator's team or a custom string input. Has built in Inputs for innocents and traitors.
- Added `ttt_logic_assigned`
- - Provides an output for the assigned team *or* role of the activator. Can be used for traitor testers.
- Added `ttt_role_button`
- - Creates a button that a certain role can use. (Hopefully not innocents. Considering hard-locking it to completely ignore 'innocent' as an input)
- - Has the ability to be locked, delayed and removed after pressed.
- - **UI is a copy of the radar, which is really shoddy and questionable. But the base of the code is there for a swap out with something better.**
- - - I can probably take a look into something cleaner, I wasn't particularly super worried about polishing it immediately.

### CODE REVIEW

**Comment by Fires1** [comment]
In future, will break separate features into separate PRs and branches.

Had to convert this over from my fork so it would have been tedious to move it all over. Now that I have access to managing branches here on this repo I'll do that going forward. 

❤️

**Comment by Fires1** [comment]
PR is good for final review

File: code/map/RoleButton/TTTRoleButton.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    [Library("ttt_role_button", Description = "Used to provide an onscreen button for a role to activate.")]
+    public partial class TTTRoleButton : Entity
+    {
+        //We don't want to touch these variables as they are the "reset" standard.
+        //Exposed to Hammer, so change them there.
+        [Property("Role", "Role which this button is exposed to. Please don't give innocents buttons.")]
+        public string Role { get; private set; } = "Traitor";
+
+        [Property("Description", "On screen tooltip shown on button")]
+        [Net]
```
[the remark this answers is not in the document]
Comment (mzegar) [reply code/map/RoleButton/TTTRoleButton.cs:18]: Not sure if there are any other alternatives for this atm.

File: code/player/TTTPlayer.RoleButtons.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                return;
+            }
+        
+            IEnumerable<TTTRoleButton> roleButtons = All.Where(x => x.GetType() == typeof(TTTRoleButton)).Select(x => x as TTTRoleButton);
+        
+            IEnumerable<TTTRoleButton> applicableButtons = roleButtons.Where(x => x.Role.ToLower() == "traitor");
+        
+            player.ClientStoreRoleButton(To.Single(ConsoleSystem.Caller), applicableButtons.Select(x => x.PackageData()).ToArray(), );
+        }
+        */
```
Comment (mzegar) [inline code/player/TTTPlayer.RoleButtons.cs:75]: I'm assuming we want to keep this in for debug purposes in the future?

File: code/teams/TTTTeam.cs [hunk 3]
```diff
@@ -36,6 +37,11 @@ public TTTTeam()
             Instance = this;
             Teams[Name] = this;
         }
+
+        public IEnumerable<Client> GetClients()
+        {
+            return Members.Select(x => x.GetClientOwner());
```
[the remark this answers is not in the document]
Comment (mzegar) [reply code/teams/TTTTeam.cs:43]: Seems pretty minor, I think we can just use both styles in our codebase.

File: code/player/TTTPlayer.RoleButtons.cs (hunk 2 above)
Comment (Fires1) [reply code/player/TTTPlayer.RoleButtons.cs:75]: Don't mind keeping it. Makes things move a lot faster when debugging

File: code/map/RoleButton/TTTRoleButton.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fires1) [reply code/map/RoleButton/TTTRoleButton.cs:18]: Not at the moment. Can be included in an issue

File: code/global/Utils.cs [hunk 4]
```diff
@@ -86,6 +87,21 @@ public static IEnumerable<Client> GetClientsSpectatingPlayer(TTTPlayer player)
             return clients;
         }
 
+        public static List<TTTPlayer> GetAlivePlayersByRole(TTTRole role)
+        {
+            List<TTTPlayer> players = new();
+
+            foreach (Client client in Client.All)
+            {
+                if (client.Pawn is TTTPlayer player && player.LifeState == LifeState.Alive && player.Role.Name == role.Name)
```
Comment (Alf21) [inline code/global/Utils.cs:96]: How about adding a "HasRole" or "IsRole" check? :)

File: code/map/TTTCreditAdjust.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    [Library("ttt_credit_adjust", Description = "Changes the amount of credits upon the activator.")]
+    public partial class TTTChangeCredits : Entity
+    {
+        [Property("Credits", "Amount of credits to remove from activator. Negative numbers add credits. Removes 1 credit by default.")]
+        public int Credits { get; set; } = 1;
+
+        [Input]
+        public void ExchangeCredits(Entity activator)
+        {
+
```
Comment (Alf21) [inline code/map/TTTCreditAdjust.cs:16]:
```suggestion
```

File: code/player/TTTPlayer.RoleButtons.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+            if (button == null)
+            {
+                Log.Warning($"Server received call for null role button with network id `{networkIdent}`.");
+                return;
+            }
+
+            if (player == null)
+            {
+                Log.Warning($"Server received call from null player to activate role button.");
```
Comment (Alf21) [inline code/player/TTTPlayer.RoleButtons.cs:92]:
```suggestion
                Log.Warning("Server received call from null player to activate role button.");
```

**Comment by Alf21** [comment]
Resolved nearly every comment. If conflicts are resolved, I gonna give it a final review

**Review by Alf21 — APPROVED** [review]

**Comment by mzegar** [comment]
This needs some more work, there are exceptions getting called on start. We probably need Fires map to finish this PR.

**Pull request merged**